src/voice/connection.rs

```diff
@@ -143,40 +165,32 @@ impl Connection {
             .next()
             .ok_or(Error::Voice(VoiceError::HostnameResolve))?;
 
-        // Important to note here: the length of the packet can be of either 4
-        // or 70 bytes. If it is 4 bytes, then we need to send a 70-byte packet
-        // to determine the IP.
-        //
-        // Past the initial 4 bytes, the packet _must_ be completely empty data.
-        //
-        // The returned packet will be a null-terminated string of the IP, and
-        // the port encoded in LE in the last two bytes of the packet.
         let udp = UdpSocket::bind("0.0.0.0:0")?;
 
+        // Follow Discord's IP Discovery procedures, in case NAT tunnelling is needed.
+        let mut bytes = [0; IpDiscoveryPacket::const_packet_size()];
         {
-            let mut bytes = [0; 70];
-
-            (&mut bytes[..]).write_u32::<BigEndian>(ready.ssrc)?;
-            udp.send_to(&bytes, destination)?;
+            let mut view = MutableIpDiscoveryPacket::new(&mut bytes[..])
+                .expect("[Voice] There are enough bytes to encode a discovery packet.");
+            view.set_pkt_type(IpDiscoveryType::Request);
+            view.set_length(70);
+            view.set_ssrc(ready.ssrc);
+        }
 
-            let mut bytes = [0; 256];
-            let (len, _addr) = udp.recv_from(&mut bytes)?;
+        udp.send_to(&bytes, destination)?;
 
-            // Find the position in the bytes that contains the first byte of 0,
-            // indicating the "end of the address".
-            let index = bytes
-                .iter()
-                .skip(4)
-                .position(|&x| x == 0)
-                .ok_or(Error::Voice(VoiceError::FindingByte))?;
+        let (len, _addr) = udp.recv_from(&mut bytes)?;
+        {
+            let view = IpDiscoveryPacket::new(&bytes[..len])
+                .expect("[voice] Should be enough space to view a packet we just fully received.");
```

Comment:
Again, if this panics there does not seem to be an issue to the user.

Reply:
This one maybe should be changed to an `Error`, since it could fail if discord completely change the API (unlike the other cases, which will only fail if the compiler is faulty).